envoy/server.go

```diff
@@ -0,0 +1,437 @@
+									// TODO: Do we need to set this? We were setting it in the REST API,
+									// but I don't think it's correct or even needed.
+									// HostRewriteSpecifier: &route.RouteAction_HostRewrite{
+									// 	HostRewrite: "",
+									// },
+								},
+							},
+						}},
+					}},
+				},
+			},
+		}
+	case "tcp":
+		connectionManagerName = wellknown.TCPProxy
+
+		connectionManager = &tcpp.TcpProxy{
+			StatPrefix: "ingress_tcp",
+			ClusterSpecifier: &tcpp.TcpProxy_Cluster{
+				Cluster: apiName,
+			},
+		}
+	default:
+		return nil, fmt.Errorf("unrecognised proxy mode: %s", svc.ProxyMode)
+	}
+
+	serialisedConnectionManager, err := ptypes.MarshalAny(connectionManager)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the connection manager: %s", err)
+	}
+
+	return &api.Listener{
+		Name: svc.Name,
+		Address: &core.Address{
+			Address: &core.Address_SocketAddress{
+				SocketAddress: &core.SocketAddress{
+					Address: s.config.BindIP,
+					PortSpecifier: &core.SocketAddress_PortValue{
+						PortValue: port,
+					},
+				},
+			},
+		},
+		FilterChains: []*listener.FilterChain{{
+			Filters: []*listener.Filter{{
+				Name: connectionManagerName,
+				ConfigType: &listener.Filter_TypedConfig{
+					TypedConfig: serialisedConnectionManager,
+				},
+			}},
+		}},
+	}, nil
+}
+
+// EnvoyClustersFromState genenerates a list of Envoy clusters from the
+// current Sidecar state
+func (s *Server) EnvoyClustersFromState() []cache.Resource {
+	s.state.RLock()
+	defer s.state.RUnlock()
+
+	// `s.state.ByService()` returns the list of service endpoints for each service.
+	// Since some services can expose multiple service ports, we need to create a
+	// separate cluster for each (service, servicePort) pair. If a service doesn't
+	// have any endpoints that are alive, we don't want to create a cluster for it.
+	//
+	// Note that in `EnvoyClustersFromState()` for the REST API we only need
+	// the first non-nil alive endpoint instance to construct the cluster
+	// because, in that case, SDS (now EDS) fetches the actual endpoints in a
+	// separate call.
+	var clusters []cache.Resource
+	clustersMap := make(map[string]*api.Cluster)
+	for svcName, svcEndpoints := range s.state.ByService() {
+		if len(svcEndpoints) < 1 {
+			continue
+		}
+
+		for _, svcEndpoint := range svcEndpoints {
+			if svcEndpoint == nil || !svcEndpoint.IsAlive() {
+				continue
+			}
+
+			for _, port := range svcEndpoint.Ports {
+				if port.ServicePort < 1 {
+					continue
+				}
+
+				envoyServiceName := SvcName(svcName, uint32(port.ServicePort))
+
+				if cluster, ok := clustersMap[envoyServiceName]; ok {
+					cluster.LoadAssignment.Endpoints[0].LbEndpoints =
+						append(cluster.LoadAssignment.Endpoints[0].LbEndpoints,
+							s.EnvoyServiceFromService(svcEndpoint, port.ServicePort)...)
+				} else {
+					envoyCluster := &api.Cluster{
+						Name: envoyServiceName,
+						// TODO: Are we happy with this timeout or should we make it configurable?
```

Comment:
We've been in production for 3 years like this so let's keep it this way until we need to configure it.